### PR TITLE
Fixed false counters resets for aggregated metrics

### PR DESCRIPTION
**Pull request by pracucci**
**What this PR does**:
I'm testing a cluster with 3K tenants per ingester and I've noticed some wrong query results when querying some of such cluster metrics. The problem is that when we internally aggregate multiple registry metrics, the order with which such values are summed is not stable and this could lead to slightly different values returned by `/metrics` which are then detected as counter resets (eg. 1st scrape from `/metrics` returns a metric with value `13304.675369502014` then 2nd scrape returns the same metric with value `13304.67536950197` even if internally the value hasn't changed).

The fix done in this PR is basically switching from `map` to `[]` (iteration order stability is not guaranteed for maps). The remaining `map` usage in `pkg/util/metrics_helper.go` should be safe.

**Which issue(s) this PR fixes**:
N/A

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by pstibrany — COMMENTED** [review]
Great find! I have some suggestions to simplify the diff in metrics_helper.go.

File: pkg/util/metrics_helper.go [hunk 1]
```diff
@@ -505,3 +491,76 @@ func (h *HistogramDataCollector) Add(hd HistogramData) {
...
 }
+
+// UserRegistry holds a Prometheus registry associated to a specific user.
+type UserRegistry struct {
+	user string
+	reg  *prometheus.Registry
+}
+
+// UserRegistries holds Prometheus registries for multiple users, guaranteeing
+// multi-thread safety and stable ordering.
```
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/util/metrics_helper.go:499]: Good point!

Discussed offline. We agreed on allowing duplicated registries here (added a comment about it) and then address this in a separate PR (either #3491 or a pre-requisite PR for #3491).

**Comment by pracucci** [comment]
Thanks @pstibrany ! I've addressed your comments.

**Review by pstibrany — APPROVED** [review]
LGTM, thanks!

**Pull request merged**